flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderRetrievalService.java

```diff
@@ -0,0 +1,178 @@
+		this.lastLeaderSessionID = null;
+
+		running = false;
+	}
+
+	@Override
+	public void start(LeaderRetrievalListener listener) {
+		checkNotNull(listener, "Listener must not be null.");
+		Preconditions.checkState(leaderListener == null, "KubernetesLeaderRetrievalService can " +
+			"only be started once.");
+
+		LOG.info("Starting {}.", this);
+
+		synchronized (lock) {
+			running = true;
+			leaderListener = listener;
+			kubernetesWatch = kubeClient.watchConfigMaps(configMapName, new ConfigMapCallbackHandlerImpl());
+		}
+	}
+
+	@Override
+	public void stop() {
+		LOG.info("Stopping {}.", this);
+
+		synchronized (lock) {
+			if (!running) {
+				return;
+			}
+			running = false;
+			if (kubernetesWatch != null) {
+				kubernetesWatch.close();
+			}
+		}
+	}
+
+	@Override
+	public String toString() {
+		return "KubernetesLeaderRetrievalService{configMapName='" + configMapName + "'}";
+	}
+
+	private class ConfigMapCallbackHandlerImpl implements FlinkKubeClient.WatchCallbackHandler<KubernetesConfigMap> {
+
+		@Override
+		public void onAdded(List<KubernetesConfigMap> configMaps) {
+			handleEvent(configMaps);
+		}
+
+		@Override
+		public void onModified(List<KubernetesConfigMap> configMaps) {
+			handleEvent(configMaps);
+		}
+
+		@Override
+		public void onDeleted(List<KubernetesConfigMap> configMaps) {
+			// Nothing to do since a new ConfigMap will be created if it is deleted externally.
+		}
+
+		@Override
+		public void onError(List<KubernetesConfigMap> configMaps) {
+			leaderListener.handleError(new Exception("Error while watching the ConfigMap " + configMapName));
+		}
+
+		@Override
+		public void handleFatalError(Throwable throwable) {
+			leaderListener.handleError(
+				new Exception("Fatal error while watching the ConfigMap " + configMapName, throwable));
+		}
+
+		private void handleEvent(List<KubernetesConfigMap> configMaps) {
+			synchronized (lock) {
+				if (running) {
+					configMaps.forEach(e -> {
+						if (e.getName().equals(configMapName)) {
+							final String leaderAddress = e.getData().get(LEADER_ADDRESS_KEY);
+							final String sessionID = e.getData().get(LEADER_SESSION_ID_KEY);
+							if (leaderAddress != null && sessionID != null) {
+								final UUID leaderSessionID = UUID.fromString(sessionID);
+								if (!(Objects.equals(leaderAddress, lastLeaderAddress) &&
+									Objects.equals(leaderSessionID, lastLeaderSessionID))) {
+									LOG.debug(
+										"New leader information: Leader={}, session ID={}.",
+										leaderAddress,
+										leaderSessionID);
+
+									lastLeaderAddress = leaderAddress;
+									lastLeaderSessionID = leaderSessionID;
+									leaderListener.notifyLeaderAddress(leaderAddress, leaderSessionID);
+								}
+							} else {
+								LOG.debug("Leader information was lost: The listener will be notified accordingly.");
+								leaderListener.notifyLeaderAddress(
+									leaderAddress, sessionID == null ? null : UUID.fromString(sessionID));
+							}
```

Comment:
I guess this logic could be factored out and shared between the Zk and K8s implementations.